feat: Incident detection via Jira PUL-1811 (#205)
codacy/pulse-user-docs: docs/one-click-integrations/jira-integration.md

# Jira integration

Pulse integrates with Jira Cloud to receive:

-   Data about **incidents**, necessary to calculate the following Accelerate metrics:

    -   [Time to recover](../metrics/accelerate.md#time-to-recover)

    -   [Change failure rate](../metrics/accelerate.md#change-failure-rate)

> -   Data bout **issues**, necessary to calculate [Lead time and Cycle time](../metrics/lead-cycle-time.md).

Reviewer: ```suggestion
-   Data about **issues**, necessary to calculate [Lead time and Cycle time](../metrics/lead-cycle-time.md).
```

Author: Thanks! Already fixed.

## Setting up the Jira integration

!!! note
    -   The Jira integration must be set up by a **Jira administrator**, as Pulse will only have access to Jira resources that the user setting up the integration has access to.
    -   You can only integrate one Pulse organization with each Jira instance.
    -   [See below](#jira-permissions) the permissions that Pulse requires from your Jira instance.

To set up the Jira integration:

1.  On Pulse, [expand **Integrations** and select **Jira**](https://app.pulse.codacy.com/integrations/jira){: target="_blank"}.

1.  Click **Install Jira App** and follow the instructions on the Jira UI to install the app. You must be a Jira administrator to proceed with this step.

    ![Installing the Pulse Jira app](images/jira-installing.png)

> 1.  Confirm that Pulse successfully created the integration and the necessary webhooks on Jira. If there was an error please [contact support](mailto:pulsesupport@codacy.com).

Reviewer: The webhook creation part is only done when 'Complete setup' is clicked. Maybe we can just remove that reference since users can't really check if webhooks were installed.

<!--TODO Add screenshot for successful installation-->

1.  Choose the strategy to detect incidents that best fits your workflows. See the [section below](#incident-detection-strategy) for a detailed description of each option.

<!--TODO Add screenshot for incident strategy-->

1.  Click **Complete setup**.

Your Jira integration is now complete. Pulse will start loading your data for the last 90 days.

<!--TODO Add screenshot for setup ok-->

## Automatic incident detection strategies {: id="incident-detection-strategy"}

The Pulse Jira integration can detect **incidents** automatically using [issues assigned with the label "Incident"](#jira-incident-label).

You can also choose [not to detect incidents via Jira](#jira-incident-not-detect).

### Use issues assigned with the label "Incident" {: id="jira-incident-label"}

> -   <!--TODO Add detection details-->

Reviewer: An issue is considered an incident if:
- It's in a terminal status (aka it's closed). In concrete Jira terms, a status of category 3 'Done' -- https://support.atlassian.com/jira-work-management/docs/workflows-and-statuses-for-the-board/
- It has the label "Incident", matched in a case insensitve manner.

If this is true, then for each value in the 'Components' field Pulse will create an incident event for the corresponding system, so values in the 'Components' field should correspond to Pulse systems (case-sensitive match).

- If a component does not already exist in Pulse as a system, no incident event will be created for it.
- If ALL components do not already exist in Pulse as systems, an incident will be created and attributed to the catch-all system `_unknown_`.

- If an issue has the "Incident" label, closing it will create an incident (or multiple, depending on the components field).
- If an issue is closed, adding the "Incident" label will create an incident (or multiple, depending on the components field).
- If an issue is closed and has the "Incident" label, adding and/or removing components will create and/or delete the corresponding incidents.
- If an issue is closed and has the "Incident" label, re-opening the issue, removing the "Incident" label or deleting the issue will delete the corresponding incidents.

Author: I updated this section. Please validate if it's ok.

### Don't detect incidents via Jira {: id="jira-incident-not-detect"}

-   Pulse doesn't detect incidents automatically using Jira events.

    Choose this option if you want to send to Pulse the information about your **incidents** using another Pulse integration - GitHub or Bitbucket integration, [PagerDuty one-click integration](pagerduty-integration.md), [Pulse CLI](../cli/cli.md), or [Ingestion API](https://ingestion.pulse.codacy.com/v1/api-docs) - or if you don't want Pulse to track incidents data.

## Collected data
<!--TODO Update this section-->

The table below lists the data that the Jira integration collects from your Jira instance, together with:

> -   The mapping between the data collected from Jira and the [Pulse data model](https://ingestion.pulse.codacy.com/v1/api-docs#tocs_event)

Author: @afsmeira The current table in this Collected data section doesn't really show the mapping between the data collected from Jira and the Pulse data model (as we have for [GitHub](https://docs.pulse.codacy.com/one-click-integrations/github-integration/#collected-data) and [Bitbucket](https://docs.pulse.codacy.com/one-click-integrations/bitbucket-integration/#collected-data)).

Can you please provide me with that mapping for Jira, including the incidents? 🙏 

I'll then move the Details into a new section and use this table for the mapping.

Reviewer: > The current table in this Collected data section doesn't really show the mapping between the data collected from Jira and the Pulse data model

Because prior to the Jira incidents feature there was no such mapping :)

After the feature we will have the following for the issues that are incidents:
- `incident_id` is the issue key (PUL-1811, for example)
- `time_created` the time the issue was created
- `time_resolved` the time the issue last transitioned to a terminal status (in concrete Jira terms, a status of category 3 'Done' -- https://support.atlassian.com/jira-work-management/docs/workflows-and-statuses-for-the-board/)
- `system` is the compoment if it already exists on Pulse as a system (but I think this limitation will probably be mentioned in a previous section).

Author: I updated the "Collected data" section according to what we talked about, excluding the detailed mapping and keeping the existing information for Lead and Cycle time. Let me know if it's ok.

-   Pulse creates new incidents when the following updates are performed in Jira:

    -   The status of an issue assigned with the label `Incident` is updated to a **Done** status. Pulse creates one or more incidents, depending on the values in the **Component(s)** field.

    -   An issue in a **Done** status is assigned with the label `Incident`. Pulse creates one or more incidents, depending on the values in the **Component(s)** field.

    -   A new component that matches an existing system is added to an issue in a **Done** status assigned with the label `Incident`. Pulse creates a new incident associated with the matching system.

-   Pulse deletes an existing incident when the following updates are performed in Jira:

    -   The corresponding issue is re-opened.

    -   The label `Incident` is removed from the corresponding issue.

    -   A component that matches an existing system is removed from the corresponding issue. Pulse deletes the incident for that system.

    -   The corresponding issue is deleted.

> !!! important
>     After completing the Jira integration setup, Pulse starts loading your incident data for the last 90 days. Therefore, **before you perform the integration setup**, make sure the corresponding Jira issues follow the rules described above, so Pulse can load your historical data correctly.
>     Pulse will only create incidents for the Jira issues in a **Done** status that are assigned with the label `Incident`, and associates them with the systems matching the values in the **Component(s)** field as described above.

Reviewer 2: This looks like it would be useful as an introduction before the instructions. I'm also adding a couple of tweak suggestions:
* present tense
* "assigned with the label" sounds odd. How about "assigned a label" or the even simpler "labeled"?

I wonder if it would make sense to move this further up before the instructions (and tweak it slightly for context):

```suggestion
!!! important
    After completing the Jira integration setup, Pulse starts loading your incident data for the last 90 days. Therefore, **before you perform the integration setup**, make sure the corresponding Jira issues follow the rules described below, so Pulse can load your historical data correctly.
    Importantly, Pulse only creates incidents for Jira issues that are in a **Done** status and labeled `Incident`, and associates them with the systems matching the values in the **Component(s)** field as described below.
```

Author: @nicklem Thank you for this suggestion. However, I'm not sure if the flow for the reader will work this way...
I need more time to check on this. I'm closing this comment for now and I'll get back to it after the launch. 👍

# Jira integration

> Pulse integrates with Jira Cloud to receive data about:
> 
> -   **Issues** - necessary to calculate [Lead time and Cycle time](../metrics/lead-cycle-time.md).
> 
> -   **Incidents** - necessary to calculate the following Accelerate metrics:

Reviewer 3: The actual objects that the integration retrieves from JIra are just issues (meaning that the "incidents" are actually also issues tagged with a special label, not a different object in Jira). So a "truer" way of saying this would be like this:

```suggestion
Pulse integrates with Jira Cloud to receive data about **issues**, necessary to calculate [Lead time and Cycle time](../metrics/lead-cycle-time.md) and the following Accelerate metrics:
```

Author: @paulo Thank you for the suggestion. 🙏  However, "incident" is the type of event that Pulse retrieves from Jira issues to calculate Time to recover and Change failure rate metrics. The documentation for the other integrations, GitHub and Bitbucket, follows the same approach. Thus, as we talked, I'll keep it as it is.

    -   [Time to recover](../metrics/accelerate.md#time-to-recover)

    -   [Change failure rate](../metrics/accelerate.md#change-failure-rate)

## Setting up the Jira integration

!!! note

>     -   The Jira integration must be set up by a **Jira administrator**, as Pulse will only have access to Jira resources that the user setting up the integration has access to.

Reviewer 3: style: Edit to use active voice. If there are some other similar sentences in the docs, it's better to just keep the consistency for now and do these improvements later in a separate pull request.

```suggestion
    -   A **Jira administrator** must set up the integration, as Pulse will only have access to Jira resources that the user setting up the integration has access to.
```

    -   You can only integrate one Pulse organization with each Jira instance.

>     -   If you want to report incidents using Jira, check the [details of the incident detection strategy](#incident-detection-strategy) before you proceed with the integration setup.

Reviewer 3: style: It's a good practice to include a verb in the link text, as it provides a clearer "call to action". Also, perhaps asking the user to **review** the details is more explicit and indicates better what we expect them to do.

```suggestion
    -   If you want to report incidents using Jira, [review the details of the incident detection strategy](#incident-detection-strategy) before you proceed with the integration setup.
```

To set up the Jira integration:

1.  On Pulse, [expand **Integrations** and select **Jira**](https://app.pulse.codacy.com/integrations/jira){: target="_blank"}.

1.  Click **Install Jira App** and follow the instructions on the Jira UI to install the app. You must be a Jira administrator to proceed with this step.

    ![Installing the Pulse Jira app](images/jira-installing.png)

1.  Confirm that Pulse successfully connected the integration to Jira. If there is an error please [contact support](mailto:pulsesupport@codacy.com).

    ![Pulse Jira App installed successfully](images/jira-installed-ok.png)

1.  Choose the strategy to detect incidents that best fits your workflows. See the [section below](#incident-detection-strategy) for a detailed description of each option.

    ![Configuring the incident detection strategy](images/jira-incident-strategy.png)

1.  Click **Complete setup**.

Your Jira integration is now complete. Pulse will start loading your data for the last 90 days.

<!--TODO Add screenshot for setup ok-->

## Automatic incident detection strategies {: id="incident-detection-strategy"}

> The Pulse Jira integration can detect **incidents** automatically using the [issues assigned with the label **Incident**](#jira-incident-label).

Reviewer 3: style: Since `Incident` is the exact label string that the users must use, I suggest the inline code formatting instead (like we already have below).

```suggestion
The Pulse Jira integration can detect **incidents** automatically using the [issues assigned with the label `Incident`](#jira-incident-label).
```

You can also choose [not to detect incidents via Jira](#jira-incident-not-detect).

### Use issues assigned with the label "Incident" {: id="jira-incident-label"}

> -   Pulse considers an incident every Jira issue in a `Done` **status** assigned with the **label** `Incident` (case-insensitive). `Done` status in Jira are all the status of an issue under the [Done category](https://support.atlassian.com/jira-work-management/docs/workflows-and-statuses-for-the-board/) (represented by the green color in Jira). Some examples are `DONE`, `CLOSED`, or `DECLINED`, but other values can be defined.

Reviewer 3: style: I suggest using a different formatting, what do you think? Bold is better to convey strings that appear on the UI, while inline code is better for user input.

If you decide to accept this formatting style, please make sure to review the other occurrences.

```suggestion
-   Pulse considers an incident every Jira issue in a **Done** status assigned with the label `Incident` (case-insensitive). **Done** status in Jira are all the status of an issue under the [Done category](https://support.atlassian.com/jira-work-management/docs/workflows-and-statuses-for-the-board/) (represented by the green color in Jira). Some examples are **DONE**, **CLOSED**, or **DECLINED**, but other values can be defined.
```

> -   Pulse associates an incident to the systems, one or more, matching the values in the `Component(s)` field of the Jira issue (case-sensitive).

Reviewer 3: edit: Small simplification, and format the field name with bold.

```suggestion
-   Pulse associates an incident to one or more systems matching the values in the **Component(s)** field of the Jira issue (case-sensitive).
```

    -   If the Jira issue is set with more than one component, Pulse creates an incident for each existing system that matches a component name, ignoring the remaining components that don't exist in Pulse as a system.

    -   If the `Component(s)` field in the Jira issue is empty, or none of the components exist in Pulse as a system, Pulse creates an incident and associates it with the system `_unknown_`.

> -   The incident creation date is the timestamp when the Jira issue was created.
> 
> -   The incident resolution date is the timestamp when the Jira issue was last updated to a `Done` status.

Reviewer 3: clean: We could merge these two items together, as they're quite related. For example:

```suggestion
-   The incident creation date is the timestamp when the Jira issue was created, while the incident resolution date is the timestamp when the Jira issue was last updated to a `Done` status.
```

    </tr>
</table>

## Which permissions does Pulse need from Jira? {: id="jira-permissions"}

Pulse requests only the necessary permissions from Jira to collect issues data from your Jira instance and [keeps your information secure](https://security.codacy.com/). See below the detailed list of permissions.

> <!--TODO Add permissions table-->

Author: @afsmeira We currently have no information regarding the permissions that Pulse needs from Jira. Do we have any internal documentation that can help me to fill in this content?
If not, can you please place here the raw information about Jira permissions, so I can work on it? 🙏

Author: @afsmeira I added the permissions information. To make it simple, I'm including only the classic scopes. Let me know if it's ok like this, or if we need to add the granular scopes also.

Reviewer: For me it's OK like this.
